rs/boundary_node/salt_sharing/canister/canister.rs

```diff
@@ -0,0 +1,163 @@
+use salt_api::{GetSaltError, GetSaltResponse, InitArg, SaltGenerationStrategy, SaltResponse};
+use std::str::FromStr;
+
+// Runs when canister is first installed
+#[init]
+fn init(init_arg: InitArg) {
+    set_timer(Duration::ZERO, || {
+        spawn(async { init_async(init_arg).await });
+    });
+}
+
+// Runs on every canister upgrade
+#[post_upgrade]
+fn post_upgrade(init_arg: InitArg) {
+    // Run the same initialization logic
+    init(init_arg);
+}
+
+#[query]
+fn get_salt() -> GetSaltResponse {
+    get_salt_response()
+}
+
+#[query(decoding_quota = 10000)]
+fn http_request(request: HttpRequest) -> HttpResponse {
+    match request.path() {
+        "/logs" => {
+            use serde_json;
+
+            let max_skip_timestamp = match request.raw_query_param("time") {
+                Some(arg) => match u64::from_str(arg) {
+                    Ok(value) => value,
+                    Err(_) => {
+                        return HttpResponseBuilder::bad_request()
+                            .with_body_and_content_length("failed to parse the 'time' parameter")
+                            .build()
+                    }
+                },
+                None => 0,
+            };
+
+            let mut entries: Log = Default::default();
+
+            for entry in export_logs(&logs::P0) {
+                entries.entries.push(LogEntry {
+                    timestamp: entry.timestamp,
+                    counter: entry.counter,
+                    priority: Priority::P0,
+                    file: entry.file.to_string(),
+                    line: entry.line,
+                    message: entry.message,
+                });
+            }
+
+            for entry in export_logs(&logs::P1) {
+                entries.entries.push(LogEntry {
+                    timestamp: entry.timestamp,
+                    counter: entry.counter,
+                    priority: Priority::P1,
+                    file: entry.file.to_string(),
+                    line: entry.line,
+                    message: entry.message,
+                });
+            }
+
+            entries
+                .entries
+                .retain(|entry| entry.timestamp >= max_skip_timestamp);
+
+            HttpResponseBuilder::ok()
+                .header("Content-Type", "application/json; charset=utf-8")
+                .with_body_and_content_length(serde_json::to_string(&entries).unwrap_or_default())
+                .build()
+        }
+        _ => HttpResponseBuilder::not_found().build(),
+    }
+}
+
+async fn init_async(init_arg: InitArg) {
+    if !is_salt_init() || init_arg.regenerate_now {
+        if let Err(err) = try_regenerate_salt().await {
+            log!(P0, "[init_regenerate_salt_failed]: {err}");
+        }
+    }
+    // Start salt generation schedule based on the argument.
+    match init_arg.salt_generation_strategy {
+        SaltGenerationStrategy::StartOfMonth => schedule_monthly_salt_generation(),
+    }
```

Comment:
here the (ad-hoc) initial generation is not separated from the scheduled generation

Reply:
I think it is, in the sense that schedule only starts once the call to `try_regenerate_salt()` returns. Or how should this separation look like?